src/effects/native/autopaneffect.cpp

```diff
@@ -0,0 +1,218 @@
+        // 1/8, 1/4, 1/2, 1, 2, 4, 8, 16, 32, 64
+        double beats = pow(2, floor(period * 9 / m_pPeriodParameter->maximum()) - 3);
+        period = groupFeatures.beat_length * beats;
+    } else {
+        // max period is 50 seconds
+        period *= sampleRate * 25.0;
+    }
+    
+    CSAMPLE stepFrac = m_pSmoothingParameter->value();
+    
+    if (gs.time > period || enableState == EffectProcessor::ENABLING) {
+        gs.time = 0;
+    }
+    
+    // Normally, the position goes from 0 to 1 linearly. Here we make steps at
+    // 0.25 and 0.75 to have the sound fully on the right or fully on the left.
+    // At the end, the "position" value can describe a sinusoid or a square
+    // curve depending of the size of those steps.
+    
+    // coef of the slope
+    // a = (y2 - y1) / (x2 - x1)
+    //       1  / ( 1 - 2 * stepfrac)
+    float a = stepFrac != 0.5f ? 1.0f / (1.0f - stepFrac * 2.0f) : 1.0f;
+    
+    // size of a segment of slope (controled by the "strength" parameter)
+    float u = (0.5f - stepFrac) / 2.0f;
+    
+    gs.frac.setRampingThreshold(positionRampingThreshold);
+    gs.frac.ramped = false;     // just for debug
+    
+    double sinusoid = 0;
+    
+    for (unsigned int i = 0; i + 1 < numSamples; i += 2) {
+        
+        CSAMPLE periodFraction = CSAMPLE(gs.time) / period;
+        
+        // current quarter in the trigonometric circle
+        float quarter = floorf(periodFraction * 4.0f);
+        
+        // part of the period fraction being a step (not in the slope)
+        CSAMPLE stepsFractionPart = floorf((quarter+1.0f)/2.0f) * stepFrac;
+        
+        // float inInterval = fmod( periodFraction, (period / 2.0) );
+        float inStepInterval = fmod(periodFraction, 0.5f);
+        
+        CSAMPLE angleFraction;
+        if (inStepInterval > u && inStepInterval < (u + stepFrac)) {
+            // at full left or full right
+            angleFraction = quarter < 2.0f ? 0.25f : 0.75f;
+        } else {
+            // in the slope (linear function)
+            angleFraction = (periodFraction - stepsFractionPart) * a;
+        }
+        
+        // transforms the angleFraction into a sinusoid.
+        // The width parameter modulates the two limits. if width values 0.5,
+        // the limits will be 0.25 and 0.75. If it's 0, it will be 0.5 and 0.5
+        // so the sound will be stuck at the center. If it values 1, the limits 
+        // will be 0 and 1 (full left and full right).
+        sinusoid = sin(M_PI * 2.0f * angleFraction) * width;
+        gs.frac.setWithRampingApplied((sinusoid + 1.0f) / 2.0f);
+        
+        pOutput[i] = pInput[i] * gs.frac * 2;
+        pOutput[i+1] = pInput[i+1] * (1.0f - gs.frac) * 2;
+        
+        gs.time++;
+    }
+    
+    // apply the delay
+//    float delay = round(m_pDelayParameter->value() * sampleRate);
+//    gs.delay->setLeftDelay( delay * sinusoid );
+    gs.delay->setLeftDelay( 0.01 * sinusoid );
```

Comment:
This does not work. I have just played with it a bit and it turns out that
```
(-0.005 * math_clamp(sinusoid, -1.0, 1.0) * sampleRate) 
```
does a nice job